Add repeater rate limiter based on aggregate response times for a domain

dannyroberts opened the pull request:
## Product Description
~No visible changes yet~ When total system use of data forwarding reaches a level that risks clogging the celery processors, in particular due to high volumes and long wait times, the data forwarding is rationed to domains (proportionally to how many users they have/pay for). This typically will result in slowing down projects that are submitting in high volume to slow endpoints.

## Technical Summary
https://dimagi.atlassian.net/browse/SAAS-15765
~This just adds the rate counters and reports them so we can get metrics, but does not actually use it yet to throttle.~
This PR originally just added rate counters and reported metrics, but now it actually implements rate limiting as well.

## Safety Assurance

### Safety story
~I'll put it up on staging to make sure the metrics and such work and it doesn't make repeaters fail. It's not being used yet to throttle so there's nothing there to test yet.~
I have tested this on staging (and if I make changes I will test them on staging), to make sure that the metrics are appearing and are behaving the way I expect, and show that throttling is also behaving the way I expect, for example respecting the thresholds I set as I change them.

### Automated test coverage

Repeaters are tested in general, but no specific testing related to this.

### QA Plan

None

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

dannyroberts committed:
Pull out delay_and_report_rate_limit as a resuable helper function

dannyroberts committed:
Add a rate counter for repeater wait milliseconds

dannyroberts committed:
Begin reporting repeat record timings to the rate counter

but do not yet actually use it to throttle

dannyroberts commented:
This is a different/complementary idea from the direction you proposed @kaapstorm. I still like yours, so this isn't really meant to supplant that

dannyroberts committed:
Fix careless error

kaapstorm reviewed (approved):
I like this approach.
* Small footprint
* Reuses existing functionality
* Easily works with changes we might make to the repeat record queue.

dannyroberts committed:
Remove stray newline

dannyroberts committed:
Fix SHOULD_RATE_LIMIT_REPEATERS defintion

dannyroberts committed:
Merge branch 'master' into dmr/repeater-rate-limiter

dannyroberts committed:
Fix units on repeater_wait thresholds (ms, not s)

dannyroberts committed:
Fix docstring in delay_and_report_rate_limit

dannyroberts committed:
Remove delaying logic from rate_limit_repeater

I realized we specifically do not want this,
because delays in high-volume celery tasks are the whole problem

dannyroberts committed:
Enable repeater rate limiting

dannyroberts committed:
Adjust repeater rate limiting to distribute retries over a 15m period

gherceg commented on corehq/project_limits/shortcuts.py:
```diff
@@ -11,3 +15,40 @@
     return STANDARD_RATIO.times(events_per_day)
+
+
+def delay_and_report_rate_limit(domain, max_wait, delay_rather_than_reject, datadog_metric, limiter):
```
What is the rationale for putting this in the `shortcuts.py` file specifically? I may just not have a good understanding of that file's purpose in the first place.

dannyroberts replied:
Oh, it just seemed like a fairly generic bit of code. Originally I reused it, but then I ended up deciding we specifically don't want delays repeaters (delays were the problem in the first place), but I liked pulling this out into something reusable so I kept it. It's technically not relevant to this PR anymore.

dannyroberts replied:
As for `shortcuts.py`, I don't know, I just didn't know a better place to put it, and this seemed like it had some reusable utilities so I used it.

gherceg replied:
Yeah my question wasn't about pulling it out into its own function, totally agree with that move. I just am unfamiliar with shortcuts.py files and if I were searching for rate limiting utilities, the shortcuts file is likely not where I'd look. All of this is moot since I'd just search for the function name anyway, but wanted to bring it up.

dannyroberts replied:
Ah, yeah. I don't know if we have a consistent naming. I think of "shortcuts" as meaning things that are part of the API of the module, but are more nice-to-have additions than core functionality. We don't have a lot of files that use that convention, just a few

gherceg commented on corehq/motech/rate_limiter.py:
```diff
@@ -0,0 +1,109 @@
+        allow_usage = True
+        delay_and_report_rate_limit(
+            domain, max_wait=max_wait, delay_rather_than_reject=delay_rather_than_reject,
+            datadog_metric='commcare.repeaters.rate_limited.test',
```
Maybe a future commit addresses this, but rename metric?

gherceg replied:
Oh I see other metrics with the same name, so maybe this is more intentional than I realized?

dannyroberts replied:
Why rename it? The `test` is intentionally part of the name, in that this is what we report when something _would_ be rate limited, but the flag is enabled to prevent that.

dannyroberts replied:
Oh yeah, I see. Yeah it's intentional

gherceg replied:
Yeah to share my perspective, my assumption when I first saw it was that it was either a temporary metric being used to test something on staging, or just a placeholder name that would be updated in a future commit. I understand the scenario that this metric is used for, and given that it is used widely won't push to rename it. Thanks.

gherceg commented on corehq/motech/rate_limiter.py:
```diff
@@ -28,7 +28,7 @@ def _get_per_user_repeater_wait_milliseconds_rate_definition(domain):
                 per_minute=0.6,
                 per_second=None,
-            ),
+            ).times(1000),
```
nit: this seems like _the way_ to set the units for a RateDefinition class, but I kinda like the idea of just passing in the units as something string constants or something (seconds, milliseconds, etc). Though again understand that this is outside of the scope of this change.

dannyroberts replied:
Yeah I thought of using a named constant like `MILLISECONDS = 1000` or something, but then I thought that was even more confusing, so I just did this.

gherceg commented on corehq/motech/repeaters/const.py:
```diff
@@ -6,6 +6,7 @@
 
 MAX_RETRY_WAIT = timedelta(days=7)
 MIN_RETRY_WAIT = timedelta(minutes=60)
+RATE_LIMITER_DELAY_RANGE = (timedelta(minutes=0), timedelta(minutes=15))
```
What is the rationale for setting the max value in this range to 15 minutes? Only asking since prior to this change, repeat records would be postponed by 60 minutes.

dannyroberts replied:
Yeah fair question. There's nothing scientific about how I chose these numbers. This just lets us guarantee that it'll be retried within 15 minutes, so there won't be major interruptions from minor overages. Not sure how necessary that is honestly, but I figured I'd try low like this and if the frequency of retrying itself (right now it'd average every 7.5m) becomes a problem we can up it.

gherceg replied:
Yeah that sounds reasonable. Thanks

dannyroberts committed:
Make RATE_LIMIT_REPEATERS a random toggle that defaults to off

dannyroberts committed:
Fix error caused by testing a random toggle without specifying namespace
